### PR TITLE
udevadm: wait: check if specified path not exist on --remove

### LINKED ISSUE #23166: Spurious fails in TEST-64-UDEV-STORAGE, season 3
**systemd version the issue has been seen with**
 > latest-ish main

**Used distribution**
 > Arch Linux

A couple of new intermittent TEST-64 issues sprouted up:

```
[   11.623012] H testsuite-64.sh[413]:   0 logical volume(s) in volume group "mdlvm_vg" now active
[   11.681957] H testsuite-64.sh[255]: + mdadm -v --stop /dev/md/mdlvm
[   11.828385] H kernel: md127: detected capacity change from 253952 to 0
[   11.828391] H kernel: md: md127 stopped.
[   11.783073] H testsuite-64.sh[416]: mdadm: stopped /dev/md/mdlvm
[   11.786673] H testsuite-64.sh[255]: + udevadm wait --settle --timeout=30 --removed /dev/md/mdlvm /dev/mdlvm_vg/mypart1 /dev/mdlvm_vg/mypart2 /dev/disk/by-id/md-name-H:mdlvm /dev/disk/by-id/md-uuid-aaaaaaaa:bbbbbbbb:ffffffff:00001010 /dev/disk/by-label/mdlvm_part
[   11.797130] H testsuite-64.sh[255]: + helper_check_device_symlinks
[   11.797798] H testsuite-64.sh[418]: + set +x
[   11.804959] H testsuite-64.sh[418]: /dev/disk/by-uuid/0317e11e-b9cc-4676-9baa-21e12ffa93ad -> /dev/sdf1
[   11.817283] H testsuite-64.sh[418]: /dev/disk/by-label/systemd_boot -> /dev/sdf1
[   11.824590] H systemd-udevd[234]: 9:127: Device is queued (SEQNUM=2306, ACTION=remove)
[   11.824977] H systemd-udevd[234]: 9:127: Device ready for processing (SEQNUM=2306, ACTION=remove)
[   11.825137] H systemd-udevd[234]: 9:127: sd-device-monitor: Passed 141 byte to netlink monitor
[   11.825244] H systemd-udevd[373]: 9:127: Processing device (SEQNUM=2306, ACTION=remove)
[   11.825350] H systemd-udevd[234]: sd-device: Failed to chase symlinks in "/sys/dev/block/9:127".
[   11.825436] H systemd-udevd[234]: Failed to create sd_device object from watch handle, ignoring: No such device
[   11.825522] H systemd-udevd[373]: 9:127: Device processed (SEQNUM=2306, ACTION=remove)
[   11.825605] H systemd-udevd[234]: md127: Device is queued (SEQNUM=2307, ACTION=remove)
[   11.825691] H systemd-udevd[234]: md127: Device ready for processing (SEQNUM=2307, ACTION=remove)
[   11.825783] H systemd-udevd[373]: 9:127: sd-device-monitor: Passed 141 byte to netlink monitor
[   11.825962] H systemd-udevd[234]: md127: sd-device-monitor: Passed 206 byte to netlink monitor
[   11.826048] H systemd-udevd[364]: md127: Processing device (SEQNUM=2307, ACTION=remove)
[   11.826137] H systemd-udevd[364]: md127: Removing watch
[   11.826241] H systemd-udevd[364]: md127: No reference left for '/dev/disk/by-id/lvm-pv-uuid-5BcdDk-USj5-rffg-jqSV-l8EZ-6AVY-5riY8i', removing
[   11.826328] H systemd-udevd[364]: md127: No reference left for '/dev/disk/by-id/md-uuid-aaaaaaaa:bbbbbbbb:ffffffff:00001010', removing
[   11.826429] H systemd-udevd[364]: md127: No reference left for '/dev/disk/by-id/md-name-H:mdlvm', removing
[   11.826512] H systemd-udevd[364]: md127: No reference left for '/dev/md/mdlvm', removing
[   11.826595] H systemd-udevd[364]: md127: Device processed (SEQNUM=2307, ACTION=remove)
[   11.826677] H systemd-udevd[364]: md127: sd-device-monitor: Passed 788 byte to netlink monitor
[   11.826759] H systemd-udevd[234]: No events are queued, removing /run/udev/queue.
[   11.826850] H systemd[1]: dev-disk-by\x2did-lvm\x2dpv\x2duuid\x2d5BcdDk\x2dUSj5\x2drffg\x2djqSV\x2dl8EZ\x2d6AVY\x2d5riY8i.device: Changed plugged -> dead
[   11.826949] H systemd[1]: dev-md-mdlvm.device: Changed plugged -> dead
[   11.827033] H systemd[1]: dev-disk-by\x2did-md\x2duuid\x2daaaaaaaa:bbbbbbbb:ffffffff:00001010.device: Changed plugged -> dead
[   11.827119] H systemd[1]: dev-disk-by\x2did-md\x2dname\x2dH:mdlvm.device: Changed plugged -> dead
[   11.827201] H systemd[1]: dev-md127.device: Changed plugged -> dead
[   11.827313] H systemd[1]: sys-devices-virtual-block-md127.device: Changed plugged -> dead
[   11.827402] H systemd[1]: sys-devices-virtual-block-md127.device: Collecting.
[   11.827486] H systemd[1]: dev-md127.device: Collecting.
[   11.827575] H systemd[1]: dev-disk-by\x2did-md\x2dname\x2dH:mdlvm.device: Collecting.
[   11.827660] H systemd[1]: dev-disk-by\x2did-md\x2duuid\x2daaaaaaaa:bbbbbbbb:ffffffff:00001010.device: Collecting.
[   11.827742] H systemd[1]: dev-md-mdlvm.device: Collecting.
[   11.827824] H systemd[1]: dev-disk-by\x2did-lvm\x2dpv\x2duuid\x2d5BcdDk\x2dUSj5\x2drffg\x2djqSV\x2dl8EZ\x2d6AVY\x2d5riY8i.device: Collecting.
[   11.827909] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2ddisk_2dby_5cx2did_2dlvm_5cx2dpv_5cx2duuid_5cx2d5BcdDk_5cx2dUSj5_5cx2drffg_5cx2djqSV_5cx2dl8EZ_5cx2d6AVY_5cx2d5riY8i_2edevice interface=org.freedesktop.DBus>
[   11.827996] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2ddisk_2dby_5cx2did_2dlvm_5cx2dpv_5cx2duuid_5cx2d5BcdDk_5cx2dUSj5_5cx2drffg_5cx2djqSV_5cx2dl8EZ_5cx2d6AVY_5cx2d5riY8i_2edevice interface=org.freedesktop.DBus>
[   11.828085] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitRemoved cookie=527 reply_cookie=0 signature=so error-name=n/a error-message=n/a
[   11.828167] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2dmd_2dmdlvm_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=528 reply_cookie=0 signature=sa{sv}as error-name=n/a error-me>
[   11.828257] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2dmd_2dmdlvm_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=529 reply_cookie=0 signature=sa{sv}as error-name=n/a error-me>
[   11.828337] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitRemoved cookie=530 reply_cookie=0 signature=so error-name=n/a error-message=n/a
[   11.828498] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2ddisk_2dby_5cx2did_2dmd_5cx2duuid_5cx2daaaaaaaa_3abbbbbbbb_3affffffff_3a00001010_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged>
[   11.828583] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2ddisk_2dby_5cx2did_2dmd_5cx2duuid_5cx2daaaaaaaa_3abbbbbbbb_3affffffff_3a00001010_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged>
[   11.828663] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitRemoved cookie=533 reply_cookie=0 signature=so error-name=n/a error-message=n/a
[   11.828743] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2ddisk_2dby_5cx2did_2dmd_5cx2dname_5cx2dH_3amdlvm_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=534 reply_cookie=0 signa>
[   11.828822] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2ddisk_2dby_5cx2did_2dmd_5cx2dname_5cx2dH_3amdlvm_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=535 reply_cookie=0 signa>
[   11.828909] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitRemoved cookie=536 reply_cookie=0 signature=so error-name=n/a error-message=n/a
[   11.828993] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2dmd127_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=537 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message>
[   11.829075] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2dmd127_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=538 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message>
[   11.829249] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitRemoved cookie=539 reply_cookie=0 signature=so error-name=n/a error-message=n/a
[   11.829355] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/sys_2ddevices_2dvirtual_2dblock_2dmd127_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=540 reply_cookie=0 signature=sa{sv}as >
[   11.829438] H testsuite-64.sh[418]: /dev/disk/by-partuuid/2e1330e2-01 -> /dev/sdf1
[   11.829629] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/sys_2ddevices_2dvirtual_2dblock_2dmd127_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=541 reply_cookie=0 signature=sa{sv}as >
[   11.829715] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitRemoved cookie=542 reply_cookie=0 signature=so error-name=n/a error-message=n/a
[   11.846595] H testsuite-64.sh[418]: /dev/disk/by-partuuid/2e1330e2-02 -> /dev/sdf2
[   11.860508] H testsuite-64.sh[418]: /dev/disk/by-id/md-uuid-aaaaaaaa:bbbbbbbb:ffffffff:00001010 -> /dev/disk/by-id/md-uuid-aaaaaaaa:bbbbbbbb:ffffffff:00001010
[   11.870683] H testsuite-64.sh[418]: ERROR: symlink '/dev/disk/by-id/md-uuid-aaaaaaaa:bbbbbbbb:ffffffff:00001010' points to '/dev/disk/by-id/md-uuid-aaaaaaaa:bbbbbbbb:ffffffff:00001010' which doesn't exist
[   11.883881] H systemd[1]: Received SIGCHLD from PID 255 (bash).
[   11.883988] H systemd[1]: Child 255 (bash) died (code=exited, status=1/FAILURE)
[   11.884069] H systemd[1]: testsuite-64.service: Child 255 belongs to testsuite-64.service.
[   11.884158] H systemd[1]: testsuite-64.service: Main process exited, code=exited, status=1/FAILURE
[   11.884270] H systemd[1]: testsuite-64.service: Failed with result 'exit-code'.
```

Full journal: [system.journal.tar.gz](https://github.com/systemd/systemd/files/8540621/system.journal.tar.gz)

```
[   12.941552] H testsuite-64.sh[255]: + mdadm -v --stop /dev/md/mdlvm
[   13.073195] H kernel: md127: detected capacity change from 253952 to 0
[   13.073203] H kernel: md: md127 stopped.
[   13.032636] H testsuite-64.sh[416]: mdadm: stopped /dev/md/mdlvm
[   13.033345] H testsuite-64.sh[255]: + udevadm wait --settle --timeout=30 --removed /dev/md/mdlvm /dev/mdlvm_vg/mypart1 /dev/mdlvm_vg/mypart2 /dev/disk/by-id/md-name-H:mdlvm /dev/disk/by-id/md-uuid-aaaaaaaa:bbbbbbbb:ffffffff:00001010 /dev/disk/by-label/mdlvm_part
[   13.042230] H testsuite-64.sh[255]: + helper_check_device_symlinks
[   13.042883] H testsuite-64.sh[418]: + set +x
[   13.050115] H testsuite-64.sh[418]: /dev/disk/by-uuid/e7a0d142-9132-49c5-bc2b-7f0a155509dc -> /dev/sdf1
[   13.062026] H testsuite-64.sh[418]: /dev/disk/by-partuuid/76293c57-01 -> /dev/sdf1
[   13.073386] H testsuite-64.sh[418]: /dev/disk/by-partuuid/76293c57-02 -> /dev/sdf2
[   13.086144] H testsuite-64.sh[418]: /dev/disk/by-label/systemd_boot -> /dev/sdf1
[   13.093016] H systemd-udevd[233]: 9:127: Device is queued (SEQNUM=2307, ACTION=remove)
[   13.093183] H systemd-udevd[233]: Validate module index
[   13.093275] H systemd-udevd[233]: Check if link configuration needs reloading.
[   13.093395] H systemd-udevd[233]: 9:127: Device ready for processing (SEQNUM=2307, ACTION=remove)
[   13.093499] H systemd-udevd[361]: 9:127: Processing device (SEQNUM=2307, ACTION=remove)
[   13.093588] H systemd-udevd[361]: 9:127: Device processed (SEQNUM=2307, ACTION=remove)
[   13.093673] H systemd-udevd[361]: 9:127: sd-device-monitor: Passed 141 byte to netlink monitor
[   13.093758] H systemd-udevd[233]: 9:127: sd-device-monitor: Passed 141 byte to netlink monitor
[   13.093842] H systemd-udevd[233]: sd-device: Failed to chase symlinks in "/sys/dev/block/9:127".
[   13.093933] H systemd-udevd[233]: Failed to create sd_device object from watch handle, ignoring: No such device
[   13.094018] H systemd-udevd[233]: md127: Device is queued (SEQNUM=2308, ACTION=remove)
[   13.094103] H systemd-udevd[233]: md127: Device ready for processing (SEQNUM=2308, ACTION=remove)
[   13.094306] H systemd-udevd[374]: md127: Processing device (SEQNUM=2308, ACTION=remove)
[   13.094475] H systemd-udevd[374]: md127: Removing watch
[   13.094595] H systemd-udevd[374]: md127: No reference left for '/dev/disk/by-id/lvm-pv-uuid-Ikzq10-Mzy4-2ZIc-zmTx-74nk-Uf0K-cWoG6G', removing
[   13.094723] H systemd-udevd[374]: md127: No reference left for '/dev/disk/by-id/md-uuid-aaaaaaaa:bbbbbbbb:ffffffff:00001010', removing
[   13.094837] H systemd-udevd[374]: md127: No reference left for '/dev/disk/by-id/md-name-H:mdlvm', removing
[   13.094945] H systemd-udevd[374]: md127: No reference left for '/dev/md/mdlvm', removing
[   13.095056] H systemd-udevd[374]: md127: Device processed (SEQNUM=2308, ACTION=remove)
[   13.095169] H systemd-udevd[374]: md127: sd-device-monitor: Passed 788 byte to netlink monitor
[   13.095402] H systemd-udevd[233]: md127: sd-device-monitor: Passed 206 byte to netlink monitor
[   13.095549] H systemd-udevd[233]: No events are queued, removing /run/udev/queue.
[   13.095663] H systemd[1]: dev-disk-by\x2did-lvm\x2dpv\x2duuid\x2dIkzq10\x2dMzy4\x2d2ZIc\x2dzmTx\x2d74nk\x2dUf0K\x2dcWoG6G.device: Changed plugged -> dead
[   13.095788] H systemd[1]: dev-md-mdlvm.device: Changed plugged -> dead
[   13.095899] H systemd[1]: dev-disk-by\x2did-md\x2duuid\x2daaaaaaaa:bbbbbbbb:ffffffff:00001010.device: Changed plugged -> dead
[   13.096012] H systemd[1]: dev-disk-by\x2did-md\x2dname\x2dH:mdlvm.device: Changed plugged -> dead
[   13.096123] H systemd[1]: dev-md127.device: Changed plugged -> dead
[   13.096230] H systemd[1]: sys-devices-virtual-block-md127.device: Changed plugged -> dead
[   13.096356] H systemd[1]: sys-devices-virtual-block-md127.device: Collecting.
[   13.096451] H systemd[1]: dev-md127.device: Collecting.
[   13.096556] H systemd[1]: dev-disk-by\x2did-md\x2dname\x2dH:mdlvm.device: Collecting.
[   13.096645] H systemd[1]: dev-disk-by\x2did-md\x2duuid\x2daaaaaaaa:bbbbbbbb:ffffffff:00001010.device: Collecting.
[   13.096737] H systemd[1]: dev-md-mdlvm.device: Collecting.
[   13.096825] H systemd[1]: dev-disk-by\x2did-lvm\x2dpv\x2duuid\x2dIkzq10\x2dMzy4\x2d2ZIc\x2dzmTx\x2d74nk\x2dUf0K\x2dcWoG6G.device: Collecting.
[   13.096914] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2ddisk_2dby_5cx2did_2dlvm_5cx2dpv_5cx2duuid_5cx2dIkzq10_5cx2dMzy4_5cx2d2ZIc_5cx2dzmTx_5cx2d74nk_5cx2dUf0K_5cx2dcWoG6G_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=561 reply_co>
[   13.097005] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2ddisk_2dby_5cx2did_2dlvm_5cx2dpv_5cx2duuid_5cx2dIkzq10_5cx2dMzy4_5cx2d2ZIc_5cx2dzmTx_5cx2d74nk_5cx2dUf0K_5cx2dcWoG6G_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=562 reply_co>
[   13.097084] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitRemoved cookie=563 reply_cookie=0 signature=so error-name=n/a error-message=n/a
[   13.097163] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2dmd_2dmdlvm_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=564 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
[   13.097243] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2dmd_2dmdlvm_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=565 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
[   13.097324] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitRemoved cookie=566 reply_cookie=0 signature=so error-name=n/a error-message=n/a
[   13.097401] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2ddisk_2dby_5cx2did_2dmd_5cx2duuid_5cx2daaaaaaaa_3abbbbbbbb_3affffffff_3a00001010_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=567 reply_cookie=0 signature=sa{sv}as error-name>
[   13.097552] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2ddisk_2dby_5cx2did_2dmd_5cx2duuid_5cx2daaaaaaaa_3abbbbbbbb_3affffffff_3a00001010_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=568 reply_cookie=0 signature=sa{sv}as error-name>
[   13.097644] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitRemoved cookie=569 reply_cookie=0 signature=so error-name=n/a error-message=n/a
[   13.097724] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2ddisk_2dby_5cx2did_2dmd_5cx2dname_5cx2dH_3amdlvm_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=570 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
[   13.097814] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2ddisk_2dby_5cx2did_2dmd_5cx2dname_5cx2dH_3amdlvm_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=571 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
[   13.097892] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitRemoved cookie=572 reply_cookie=0 signature=so error-name=n/a error-message=n/a
[   13.097970] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2dmd127_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=573 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
[   13.098049] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/dev_2dmd127_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=574 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
[   13.098125] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitRemoved cookie=575 reply_cookie=0 signature=so error-name=n/a error-message=n/a
[   13.098202] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/sys_2ddevices_2dvirtual_2dblock_2dmd127_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=576 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
[   13.098281] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/sys_2ddevices_2dvirtual_2dblock_2dmd127_2edevice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=577 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
[   13.098358] H systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitRemoved cookie=578 reply_cookie=0 signature=so error-name=n/a error-message=n/a
[   13.099030] H testsuite-64.sh[418]: /dev/disk/by-diskseq/4 -> /dev/sdd
[   13.110605] H testsuite-64.sh[418]: /dev/disk/by-diskseq/1 -> /dev/sda
[   13.122368] H testsuite-64.sh[418]: /dev/disk/by-diskseq/3 -> /dev/sdc
[   13.133983] H testsuite-64.sh[418]: /dev/disk/by-diskseq/2 -> /dev/sdb
[   13.145397] H testsuite-64.sh[418]: /dev/disk/by-diskseq/5 -> /dev/sde
[   13.157097] H testsuite-64.sh[418]: /dev/disk/by-diskseq/15 -> /dev/sdf
[   13.173040] H testsuite-64.sh[418]: /dev/disk/by-path/pci-0000:00:03.0-ata-4 -> /dev/sdd
[   13.189922] H testsuite-64.sh[418]: /dev/disk/by-path/pci-0000:00:03.0-ata-4.0 -> /dev/sdd
[   13.206842] H testsuite-64.sh[418]: /dev/disk/by-path/pci-0000:00:03.0-ata-2.0 -> /dev/sda
[   13.223731] H testsuite-64.sh[418]: /dev/disk/by-path/pci-0000:00:03.0-ata-2 -> /dev/sda
[   13.240622] H testsuite-64.sh[418]: /dev/disk/by-path/pci-0000:00:03.0-ata-3.0 -> /dev/sdc
[   13.257988] H testsuite-64.sh[418]: /dev/disk/by-path/pci-0000:00:03.0-ata-3 -> /dev/sdc
[   13.269692] H testsuite-64.sh[418]: /dev/disk/by-path/pci-0000:00:03.0-ata-1.0 -> /dev/sdb
[   13.287866] H testsuite-64.sh[418]: /dev/disk/by-path/pci-0000:00:03.0-ata-1 -> /dev/sdb
[   13.299980] H testsuite-64.sh[418]: /dev/disk/by-path/pci-0000:00:01.1-ata-1.0-part1 -> /dev/sdf1
[   13.317753] H testsuite-64.sh[418]: /dev/disk/by-path/pci-0000:00:01.1-ata-1-part1 -> /dev/sdf1
[   13.335931] H testsuite-64.sh[418]: /dev/disk/by-path/pci-0000:00:01.1-ata-1.0-part2 -> /dev/sdf2
[   13.354175] H testsuite-64.sh[418]: /dev/disk/by-path/pci-0000:00:01.1-ata-1-part2 -> /dev/sdf2
[   13.372319] H testsuite-64.sh[418]: /dev/disk/by-path/pci-0000:00:03.0-ata-5.0 -> /dev/sde
[   13.392629] H testsuite-64.sh[418]: /dev/disk/by-path/pci-0000:00:03.0-ata-5 -> /dev/sde
[   13.413682] H testsuite-64.sh[418]: /dev/disk/by-path/pci-0000:00:01.1-ata-1 -> /dev/sdf
[   13.426748] H testsuite-64.sh[418]: /dev/disk/by-path/pci-0000:00:01.1-ata-1.0 -> /dev/sdf
[   13.445624] H testsuite-64.sh[418]: /dev/disk/by-id/md-name-H:mdlvm -> /dev/disk/by-id/md-name-H:mdlvm
[   13.454370] H testsuite-64.sh[418]: ERROR: symlink '/dev/disk/by-id/md-name-H:mdlvm' points to '/dev/disk/by-id/md-name-H:mdlvm' which doesn't exist
[   13.466326] H systemd[1]: systemd-journald.service: Received EPOLLHUP on stored fd 59 (stored), closing.
[   13.466618] H systemd[1]: Received SIGCHLD from PID 255 (bash).
[   13.466714] H systemd[1]: Child 255 (bash) died (code=exited, status=1/FAILURE)
```

Full journal: [system.journal.tar.gz](https://github.com/systemd/systemd/files/8540632/system.journal.tar.gz)

It appears to be essentially the same issue, but it's weird the offending (and aparently dangling) symlink exists, when we call `udevadm wait --settle --remove ...` on it beforehand.

/cc @yuwata

### PR DESCRIPTION
**Pull request by yuwata**
Even if the corresponding device node or syspath are already removed,
the specified symlink to the device node may still exist.

Fixes #23166.

### CODE REVIEW

**Comment by yuwata** [comment]
cc @mrc0mmand

**Review by mrc0mmand — APPROVED** [review]
Nice, thanks for the quick fix!

File: src/udev/udevadm-wait.c [hunk 1]
```diff
@@ -48,22 +49,25 @@ static int check_device(const char *path) {
...
         assert(path);
 
+        if (arg_wait_until == WAIT_UNTIL_REMOVED) {
+                r = laccess(path, F_OK);
+                if (r == -ENOENT)
+                        return true;
+                if (r < 0)
+                        return r;
+                return false;
+        }
```
Comment (poettering) [inline src/udev/udevadm-wait.c:59]: hmm, I don't understand this change? why do we need to handle this beforehand explicitly, can you elaborate?

I mean, if I point sd_device_new_from_path() to a dangling symlink or on an absent device should not matter?

i.e. sd_device_new_from_path() does "stat()" on the passed path, and if that results in ENOENT (as a dangling symlink) would, then the function will return ENODEV, which the previous code was already checking for?

**Comment by poettering** [comment]
or in other words, this has a certain TTOCTTOU feel to it... why do we need to check in advance what sd-device_new_from_path() should be able to figure out on its own? Can you elaborate?

**Comment by yuwata** [comment]
@poettering The issue here is, when a user specified a symlink to a device node, the previous logic may allow `udevadm wait --removed` being finished even though the symlink still exists, though in that case the symlink is broken or dangling. Why? When `check_device()` is called, the device node and/or the corresponding syspath may be already removed (hence, `sd_device_new_from_path()` fails with -ENODEV), but udevd may not processed the 'remove' uevent yet.

On `--removed`, checking the specified path with `laccess()` ensures the symlink not exist anymore. Of course, still there may be a chance that udevd is still processing the corresponding 'remove' uevent, but we can avoid the chance by setting `--settle` option.

**Pull request merged**